### PR TITLE
Generalized DCI generation

### LINKED ISSUE #679: Generalize tiling support in DCIs - Use stride_map
### What feature would you like to be added?

Current DCI generation (host to device and device to host transfers) support tiling the stick dimension only ie `dim_map`s such as `[1,2,0,2]` where 2 is the stick dimension and `device_size` is `[.. , 64]`.

**Update**: with the introduction of the `stride_map` in the `SpyreTensorLayout` the solution to this problem will be different from the mock code below and more general.

### Why is this feature needed?

We need to generalize the current implementation to support tiling arbitrary dimensions an arbitrary number of times and in arbitrary order to support view operations such as fusing non-stick dimensions.

### Any alternative solution that you may want to share?

Here is a python implementation of DCIs with general tiling (but not accounting for padding and only fp16) in python:
```python
# work around limitations of to_with_layout
def to_with_layout_fixed(t, stl):
    n = len(t.size())

    # relayout on host
    r = torch.empty(stl.device_size, dtype=torch.float16)
    for i in itertools.product(*map(range, stl.device_size)):
        coordinates = [0] * n
        strides = [1] * n
        for j in range(len(stl.dim_map) - 1, -1, -1):
            k = stl.dim_map[j]
            coordinates[k] += i[j] * strides[k]
            strides[k] *= stl.device_size[j]
        r[*i] = t[*coordinates]
    r = r.reshape(t.size())

    # send to device without tiling or reordering dimensions using row major layout
    row_major_device_size = list(t.size())[:-1] + [t.size()[-1] // 64, 64]
    row_major_dim_map = list(range(n)) + [n - 1]
    r = to_with_layout(
        r,
        SpyreTensorLayout(
            row_major_device_size, row_major_dim_map, DataFormats.SEN169_FP16
        ),
    )

    # restore intended layout
    set_spyre_tensor_layout(r, stl)

    return r
```

### Anything else we need to know?

The current implementation does not handle stride computations properly when tiling non-stick dimensions or tiling the stick dimension more than once.

### LINKED ISSUE #896: Add DCI support for views
### What feature would you like to be added?

When creating the DataConversionInfo DCI for a CPU tensor that is a view (e.g., CPU strides reflect a transposition), the DCI should support transposing this tensor during the data conversion step on the host through the strides.

### Why is this feature needed?

To support transferring view tensors to and from the device.

### Any alternative solution that you may want to share?

_No response_

### Anything else we need to know?

Code to reproduce the error:
```
import torch
DEVICE = torch.device("spyre")

x = torch.randn((128, 64), dtype=torch.float16)
x = x.t()

x_device = x.to(DEVICE)

cpu_delta = torch.abs(x - x_device.cpu()).max()
print(f"Max delta Spyre vs. CPU: {cpu_delta}")
```

### LINKED ISSUE #899: Support padding non-stick dimensions in DCI
### What feature would you like to be added?

DCI support for tensor with padding in non-stick dimensions. This enables padding to be added and removed when tensors are transferred. 

### Why is this feature needed?

To support padding tensors in dimensions other than the stick dimension.

### Any alternative solution that you may want to share?

_No response_

### Anything else we need to know?

To handle this in the DCI, the size `data_conversion_stride_info` of a padded non-stick dimension in the should be the non-padded/cpu size.

### PR DESCRIPTION
**Pull request by willjones1989**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Generalizes the DCI generation using the `stride_map` to support:

- Views
- Folding
- Expansions
- Slices (storage offset > 0)
- Padding of arbitrary dimensions
- Tiling of arbitrary dimensions

#### Which issue(s) this PR is related to:

Fixes #679
Fixes #896
Fixes #899

### CODE REVIEW

**Review by tardieu — DISMISSED** [review]
lgtm

File: tests/tensor/test_tensor_layout.py [hunk 1]
```diff
@@ -119,16 +176,4727 @@ def test_to_spyre_layout(self):
...
+            ([40, 120], [120, 1], [1, 120, 64], [0, 1, 0], [7680, 1, 120]),
+            ([40, 120], [120, 1], [2, 40, 64], [1, 0, 1], [64, 120, 1]),
+            ([40, 128], [128, 1], [1, 128, 64], [0, 1, 0], [8192, 1, 128]),
+            ([40, 128], [128, 1], [2, 40, 64], [1, 0, 1], [64, 128, 1]),
+            ([40, 240], [240, 1], [1, 240, 64], [0, 1, 0], [15360, 1, 240]),
+            ([40, 240], [240, 1], [4, 40, 64], [1, 0, 1], [64, 240, 1]),
+            ([40, 256], [256, 1], [1, 256, 64], [0, 1, 0], [16384, 1, 256]),
+            ([40, 256], [256, 1], [4, 40, 64], [1, 0, 1], [64, 256, 1]),
+            # (
+            #     [20, 40, 60],
```
Comment (thoangtrvn) [inline tests/tensor/test_tensor_layout.py]: Do we need to keep these commented code?

Comment (tardieu) [reply tests/tensor/test_tensor_layout.py]: These tests are expensive to always run but we should keep them to be enabled once we have tiering support.

File: torch_spyre/csrc/spyre_mem.cpp [hunk 2]
```diff
@@ -61,189 +62,306 @@ struct DMAParameters {
...
-        pad_stride_info.size_[0] = cpu_shape[stl.dim_map.back()] % stick_size;
-        break;
+        dcsi_sizes[tile_index] = std::min(remaining_elements, tile_size);
+
+        elements_before *= dcsi_sizes[tile_index];
+      } else {
+        // When the current elements is not evenly divisible by the tile stride
+        // then this tile and the next tile have a remainder.
+        //
+        // In thse cases we get both tile and compute the dcsi sizes and
```
Comment (thoangtrvn) [inline torch_spyre/csrc/spyre_mem.cpp]: typo `thse` -> `these`.

File: tests/tensor/test_tensor_layout.py (hunk 1 above)
Comment (thoangtrvn) [reply tests/tensor/test_tensor_layout.py]: @tardieu : can we put into a dedicated test function, and set it to be skipped, 

```
@pytest.mark.slow 
def test_stl_copyable_expensive_cases():
   ...
```

then update `pytest_models.ini` with 
```
...
     slow: marks test as slow (skipped by default)

addopts = -m  "not bf16operation and not slow"
```

This allow users/CICD to run it on demand as needed. 
```sh
python -m pytest -m "slow" tests/ 
```

**Review by thoangtrvn — CHANGES_REQUESTED** [review]
I think we can we move the expensive test cases to a dedicated ones, it can be in a separated test files with heavy test cases for easier tracking.  The tests remain visible and documented in the codebase.

File: torch_spyre/csrc/spyre_mem.cpp [hunk 3]
```diff
@@ -61,189 +62,306 @@ struct DMAParameters {
...
- * the dimension size divided by the stick size (rounded down). The cpu offset
- * is the dimension size divided by the stick size (rounded up), multiplied by
- * the stick size. The device offset is the size of the stick size multiplied by
- * the volume of the dimensions preceeding the stick dim on the device.
+ * Fills out size and strides for each dimension of the tensor.
  *
  * @param sizes: dimension sizes of the CPU tensor
  * @param strides: dimension strides of the CPU tensor
- * @param dev_shape: shape of tensor on device
+ * @param storage_offset: stoage offset of the CPU tensor
```
Comment (JRosenkranz) [inline torch_spyre/csrc/spyre_mem.cpp]: typo: stoage -> storage

File: tests/tensor/test_tensor_layout.py (hunk 1 above)
Comment (willjones1989) [reply tests/tensor/test_tensor_layout.py]: I have removed the commented tests to avoid delaying this PR.
They can be added back in subsequent PR that updates the test configuration files.

**Review by thoangtrvn — APPROVED** [review]
LGTM. 
This change enables the Spyre backend to handle more complex tensor layouts and operations.

**Review by tardieu — APPROVED** [review]
lgtm

**Comment by ani300** [comment]
~DCO is now enabled for the repo, you will need to re-do all your commits with signature so we can merge~

Something else is going on?

**Comment by tardieu** [comment]
Unverified commits.
